extensions/amp-story/0.1/amp-story-info-dialog.js

```diff
@@ -0,0 +1,252 @@
+
+/** @const {string} Class to toggle the info dialog. */
+export const DIALOG_VISIBLE_CLASS = 'i-amphtml-story-info-dialog-visible';
+
+/** @const {string} Class to toggle the info dialog link. */
+export const MOREINFO_VISIBLE_CLASS = 'i-amphtml-story-info-moreinfo-visible';
+
+/**
+ * A dialog that provides a link to the canonical URL of the story, as well as
+ * a link to any more information that the viewer would like to provide about
+ * linking on that platform.
+ */
+export class InfoDialog {
+  /**
+   * @param {!Window} win
+   * @param {!Element} parentEl Element where to append the component
+   */
+  constructor(win, parentEl) {
+    /** @private @const {!Window} */
+    this.win_ = win;
+
+    /** @private {?Element} */
+    this.element_ = null;
+
+    /** @private {?Element} */
+    this.innerContainerEl_ = null;
+
+    /** @private {boolean} */
+    this.isBuilt_ = false;
+
+    /** @private @const {!./localization.LocalizationService} */
+    this.localizationService_ = Services.localizationServiceV01(this.win_);
+
+    /** @private @const {!./amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = Services.storyStoreServiceV01(this.win_);
+
+    /** @private @const {!Element} */
+    this.parentEl_ = parentEl;
+
+    /** @private @const {!../../../src/service/resources-impl.Resources} */
+    this.resources_ = Services.resourcesForDoc(getAmpdoc(this.win_.document));
+
+    /** @private {?Element} */
+    this.moreInfoLinkEl_ = null;
+
+    /** @const @private {!../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = Services.viewerForDoc(this.parentEl_);
+  }
+
+  /**
+   * Builds and appends the component in the story.
+   * @return {!Promise}
+   */
+  build() {
+    if (this.isBuilt()) {
+      return Promise.resolve();
+    }
+
+    this.isBuilt_ = true;
+    const root = this.win_.document.createElement('div');
+    const html = htmlFor(this.parentEl_);
+    this.element_ = html`
+        <div class="i-amphtml-story-info-dialog i-amphtml-story-system-reset">
+          <div class="i-amphtml-story-info-dialog-container">
+            <h1 class="i-amphtml-story-info-heading"></h1>
+            <a class="i-amphtml-story-info-link"></a>
+            <a class="i-amphtml-story-info-moreinfo"></a>
+          </div>
+        </div>`;
+
+    createShadowRootWithStyle(root, this.element_, CSS);
+    this.initializeListeners_();
+
+    this.innerContainerEl_ = this.element_
+        .querySelector('.i-amphtml-story-info-dialog-container');
+
+    const appendPromise = this.resources_.mutateElement(this.parentEl_, () => {
+      this.parentEl_.appendChild(root);
+    });
+
+    const pageUrl = Services.documentInfoForDoc(
+        getAmpdoc(this.parentEl_)).canonicalUrl;
+
+    return Promise.all([
+      appendPromise,
+      this.setHeading_(),
+      this.setPageLink_(pageUrl),
+      this.requestMoreInfoLink_()
+          .then(moreInfoUrl => this.setMoreInfoLinkUrl_(moreInfoUrl)),
```

Comment:
Don't know if you want to wait on this promise since it could take a long time if viewer times out.

Reply:
The promise is actually only used for testing; not sure how best to denote that